### ISSUE TITLE
[MU4 Task] Implement fuzzy text search

### ISSUE THREAD

**Issue by shoogle** [musescore/MuseScore#11411]
**Task description**
Allow synonyms or inexact matches when performing a text search. This could allow the "Voice" instrument to be included in results of a search for "vocal", for example.

Places where a fuzzy text search could be useful:

- Palette search
- Instruments
- Templates
- Plugins
- Recent scores
- etc.

This feature was requested by an accessibility tester. It's particularly useful for low vision users who may not notice that search results are being narrowed as they type.

<img width="292" alt="image" src="https://user-images.githubusercontent.com/11011881/165414912-8585c624-e8a1-485a-bfee-1c1ad343ccc5.png">

<img width="291" alt="image" src="https://user-images.githubusercontent.com/11011881/165414998-05a97428-51fa-41df-b88a-5f37f1d402be.png">

A user with low vision might finish typing "vocal" before checking the results and wondering why there are no matches, whereas a sighted user would probably stop typing after just "vo" because that is already enough to return the "Voice" instrument.

**Comment by wizofaus** [comment]
Palette search isn't even just a matter of being "fuzzy", but for instance none of the following searches find anything:

"ottava"
"D.C."  (but "D.S." works!)
"Dal Segno" (but Da Capo works!)
"forte"
"decresc"
"riten"
"ritard"
"molto"/"poco"  (surprisingly there *are* no preset palette commands that have either term in them)
"portamento"
"1st"
etc.

Others like "sp" don't find the expected palette command (there is no ***sp*** dynamic by default I guess?)

**Comment by raphj** [comment]
@wizofaus I have the same issue with "forte", "decr", and also "meso" which I expect to find "mf".
For descr, this is a **regression**, at least in the French translation: MuseScore 3 finds decrescendo with "descr" because if the element was already called diminuendo back then, it had "(descrescendo)" in its name so the search found it. 

It seems like a separate issue which does not need fuzzy search to be fixed, maybe just some careful labeling work. Should we open a separate issue? @shoogle?

**Comment by shoogle** [comment]
@raphj, you're right. This issue is really about recognising synonyms. Feel free to open a different one about fuzzy text search, which is more to do with typos in search text rather than alternative names for the same thing.

**Comment by raphj** [comment]
@shoogle I'm not too concerned about fuzzy search myself, though it would be a nice feature so here you go: https://github.com/musescore/MuseScore/issues/15983

However, I'm quite interested in the present one. Do you have something planned for it? I thought about adding context / labeling in the names of musical items in parentheses, would that be appreciated? Like "Diminuendo" → "Diminuendo (decrescendo)", or "f" → "f (forte)". But I guess if these were removed there might be a reason?

Or maybe another solution could be adding hidden (or not)  text data / tags so the search engine can find musical items using those things?

**Comment by NEVED123** [comment]
In addition, you should also be able to type "sharp" or "#" to search for palette items with a ♯ (U+266F) in their name (i.e search "F# major" comes up with no results, but "F♯ major" does).

**Comment by zacjansheski** [comment]
As another example to "add to the list" here: https://github.com/musescore/MuseScore/issues/21402#issuecomment-2184157447

Searching for Drumset, Drum Set, Drum Kit, Traps, Trap kit should all produce the drum kit instruments.  Especially from a percussion perspective, many instruments have accepted alternative names and spellings.
